### PR TITLE
[PatternMatch] Introduce m_c_Select

### PR DESCRIPTION
**Pull request by davemgreen**
This matches m_Select(C, L, R) or m_Select(C, R, L).

### CODE REVIEW

File: llvm/include/llvm/IR/PatternMatch.h [hunk 1]
```diff
@@ -1771,6 +1776,13 @@ m_SelectCst(const Cond &C) {
   return m_Select(C, m_ConstantInt<L>(), m_ConstantInt<R>());
 }
 
+/// Match Select(C, LHS, RHS) or Select(C, RHS, LHS)
+template <typename Cond, typename LHS, typename RHS>
+inline ThreeOps_match<Cond, LHS, RHS, Instruction::Select, true>
+m_c_Select(const Cond &C, const LHS &L, const RHS &R) {
+  return ThreeOps_match<Cond, LHS, RHS, Instruction::Select, true>(C, L, R);
```
Comment (arsenm) [inline llvm/include/llvm/IR/PatternMatch.h:1783]: I think this needs to hide the condition, or add something to indicate the condition needs to be inverted. You can't safely directly use the condition if it's swapped

**Comment by dtcxzyw** [comment]
Reverse ping.

**Comment by dtcxzyw** [comment]
>  I'm not sure if we considered it worth adding or not.

I think it is useful to avoid duplicate code when the condition is negligible.

**Comment by davemgreen** [comment]
I've removed the first condition. My C++ is probably worse now than when I started working on a compiler, so let me know if anything looks wrong :)

**Review by dtcxzyw — APPROVED** [review]
LGTM. Thank you!

**Review by nikic — APPROVED** [review]

**Pull request merged**